### PR TITLE
[New] `no-unresolved`: add `caseSensitiveStrict` option

### LINKED ISSUE #1259: Incorrect file resolution for paths higher than CWD
### Env:
Windows 7 x64, NodeJs v10.15.0, eslint 5.11.1, eslint-plugin-import: 2.14.0

### Repro steps: 
Rule **no-unresolved** is turned on with **case sensitive** option
``` javascript
"rules": {
        "import/no-unresolved": ["error", { caseSensitive: true}]
 }
```

Project root is **"B/"**
CWD is **"C/"**

Project Structure (tree)
```
├─ A
   ├─ B
   │  ├─ C
   │  │  ├─ D.js
```

``` javascript
import D from 'B/C/D.js'; // should be no error (casing is preserved correctly)
```

``` javascript
import D from 'B/C/d.js'; // causes error (filename is misspelled)
```

``` javascript
import D from 'B/c/D.js'; // causes error (cwd is misspelled)
```

**Expected result:**
``` javascript
import D from 'b/C/D.js'; // should cause error (project root is misspelled)
```

**Actual result:**
``` javascript
import D from 'b/C/D.js'; // no error (project root is misspelled)
```

### Description:
Eslint task is launched by grunt, where grunt config file is located in _**"C/"**_ folder (CWD). While actual project folder is one ore more levels higher than cwd (_**"B/"**_ ). So imports with such path don't cause error.

It is caused by  _"fileExistsWithCaseSync"_ function
https://github.com/benmosher/eslint-plugin-import/blob/1cd82eb27df85768fbd076e4ff6b7f36d6f652ce/utils/resolve.js#L38
So it basically doesn't go any higher than cwd to cache path.

**Suggestion:**
It either should not be limited at all (because it is relatively cheap operation as it caches previous paths), 
``` javascript
import D from 'a/B/C/D.js'; // should also cause error (it is higher than project root)
```
or at least it should be limited to project dir (something like getBaseDir() is doing - line 153)
``` javascript
import D from 'b/C/D.js'; // should cause error (project root is misspelled)
```

### PR DESCRIPTION
**Pull request by sergei-startsev**
Fix #1259 issue. The code above should be identified as invalid by `import/no-unresolved` rule with enabled `caseSensitiveStrict` option:

```js
/*eslint import/no-unresolved: [2, { caseSensitiveStrict: true }]*/

// Absolute paths
import Foo from `/Users/fOo/bar/file.js` // reported, /Users/foo/bar/file.js
import Foo from `d:/fOo/bar/file.js` // reported, d:/foo/bar/file.js

// Relative paths, cwd is Users/foo/
import Foo from `./../fOo/bar/file.js` // reported
```

### CODE REVIEW

**Comment by sergei-startsev** [comment]
@benmosher @ljharb Have a chance to take a look? The changes are behind the option, no breaking changes are expected.

**Comment by benmosher** [comment]
Ah interesting, if just plain `caseSensitive` is not properly checking paths with relative bits, I think that is a bug and a new option is not necessary. @ljharb can check my math on that though. 

But my recommendation would be to make `caseSensitive` report on your case as well.

**Comment by benmosher** [comment]
Feels like any explicit path segment should be checked, regardless of the CWD. Not sure how exactly that would work though.

**Comment by sergei-startsev** [comment]
@benmosher right, that's how `caseSensitive` worked initially before ignoring `cwd` part.

**Comment by ljharb** [comment]
@sergei-startsev sorry for the long delay here. if you'd mind rebasing, i'll be happy to take a fresh look.

**Comment by sergei-startsev** [comment]
@ljharb The PR has been rebased, take a look.

**Comment by ljharb** [comment]
@sergei-startsev what happens if someone sets `caseSensitive` to `false`, but `caseSensitiveStrict` to true? Same question if both are true?

**Comment by sergei-startsev** [comment]
@ljharb Good point, rethought `caseSensitiveStrict`, it'll win in both cases now.

**Review by ljharb — APPROVED** [review]
Thanks!

**Comment by sergei-startsev** [comment]
@ljharb Do you plan to merge the PR yourself?

**Comment by ljharb** [comment]
@sergei-startsev yep! i'm holding off because i want to find a fix for #2199, so i can release a v2.24 patch release, before merging this in for a v2.25.

**Comment by ljharb** [comment]
@sergei-startsev travis should be working again; i've rebased this branch so you can see the logs.

**Comment by sergei-startsev** [comment]
@ljharb Not sure what went wrong last time, but the checks are green on all platforms now (thanks for fixing travis and adding appveyor). The commit with logs was reverted (not sure why codecov hasn't rerun).

**Pull request merged**
